src/block/tile/util/ContainerHelper.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\block\tile\util;
+
+use pocketmine\block\tile\Container;
+use pocketmine\data\bedrock\item\SavedItemData;
+use pocketmine\data\bedrock\item\SavedItemStackData;
+use pocketmine\data\SavedDataLoadingException;
+use pocketmine\item\Item;
+use pocketmine\nbt\NBT;
+use pocketmine\nbt\tag\CompoundTag;
+use pocketmine\nbt\tag\ListTag;
+
+final class ContainerHelper{
+
+	private function __construct(){
+		//NOOP
+	}
+
+	/**
+	 * @return Item[]|null
+	 * @phpstan-return array<int, Item>|null
+	 */
+	public static function deserializeContents(CompoundTag $tag) : ?array{
+		if(($inventoryTag = $tag->getTag(Container::TAG_ITEMS)) instanceof ListTag && $inventoryTag->getTagType() === NBT::TAG_Compound){
+			$contents = [];
+			/** @var CompoundTag $itemNBT */
+			foreach($inventoryTag as $slot => $itemNBT){
+				try{
+					$count = $itemNBT->getByte(SavedItemStackData::TAG_COUNT);
+					if($count === 0){
+						continue;
+					}
+					$contents[$slot] = Item::nbtDeserialize($itemNBT);
```

Comment:
IIRC, this will break loading container data from vanilla worlds because the save format in Chiseled Bookshelves and generic containers is a little bit different. Chiseled Bookshelves have their slot stored as a key in inventory compound tag (how it’s implemented here), but generic containers save their slot in `SavedItemStackData::TAG_SLOT` byte tag. See https://github.com/pmmp/PocketMine-MP/blob/stable/src/block/tile/ContainerTrait.php#L55

Reply:
I didn't notice that detail. Thanks for pointing it out. I'll just make chiseled bookshelves have their own code